<a href="https://colab.research.google.com/github/aparna-2001/cleveland-health-data-matrix-analysis/blob/main/PCA_cleveland_health_data_matrix_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import csv
import zipfile
import io


zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

try:
    with zipfile.ZipFile(zip_file_path, 'r') as zf:
        print(f"Files found inside '{zip_file_path}': {zf.namelist()}\n")


        if csv_filename_in_zip in zf.namelist():
            with zf.open(csv_filename_in_zip) as csv_file:
                # zf.open returns a file-like object in binary mode by default.
                # csv.reader expects a text file, so we wrap it in io.TextIOWrapper
                text_stream = io.TextIOWrapper(csv_file, encoding='utf-8') # Or the correct encoding of your CSV
                csv_rows = list(csv.reader(text_stream))
                print(f"Successfully read {len(csv_rows)} rows from '{csv_filename_in_zip}'.")

                # Printing the first few rows for demonstration
                if csv_rows:
                    print("\nFirst 5 rows:")
                    for i, row in enumerate(csv_rows[:5]):
                        print(row)
                    if len(csv_rows) > 5:
                        print("...")
                else:
                    print("The CSV file is empty.")
        else:
            print(f"Error: The CSV file '{csv_filename_in_zip}' was not found inside '{zip_file_path}'.")
            print(f"Please choose one of the files listed above.")

except zipfile.BadZipFile:
    print(f"Error: '{zip_file_path}' is not a valid zip file.")
except FileNotFoundError:
    print(f"Error: The file '{zip_file_path}' was not found. Please ensure it is uploaded or the path is correct.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")


Files found inside 'archive.zip': ['heart_cleveland_upload.csv']

Successfully read 298 rows from 'heart_cleveland_upload.csv'.

First 5 rows:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'condition']
['69', '1', '0', '160', '234', '1', '2', '131', '0', '0.1', '1', '1', '0', '0']
['69', '0', '0', '140', '239', '0', '0', '151', '0', '1.8', '0', '2', '0', '0']
['66', '0', '0', '150', '226', '0', '0', '114', '0', '2.6', '2', '0', '0', '0']
['65', '1', '0', '138', '282', '1', '2', '174', '0', '1.4', '1', '1', '0', '1']
...


In [7]:
import csv
import numpy as np
import zipfile
import pandas as pd

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

#opening the csv from zipfile

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:

#conversion of the csv to table

          data_table = pd.read_csv(file)
          #print(data_table)


#describing the data

print(data_table.columns.to_list(),'\n')
print(data_table.head(),'\n')
print(data_table.info(),'\n')
print(data_table.describe(),'\n')



['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'condition'] 

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   69    1   0       160   234    1        2      131      0      0.1      1   
1   69    0   0       140   239    0        0      151      0      1.8      0   
2   66    0   0       150   226    0        0      114      0      2.6      2   
3   65    1   0       138   282    1        2      174      0      1.4      1   
4   64    1   0       110   211    0        2      144      1      1.8      1   

   ca  thal  condition  
0   1     0          0  
1   2     0          0  
2   0     0          0  
3   1     0          1  
4   0     0          0   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 297 entries, 0 to 296
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   age        297 non-null    int64  
 1   sex

**Inference**



*   The data contains 297 entries with 14 columns
*   The mean age is 54.542088,
           - which mean most of the recorded patients ages are clustered around 54

*   In the 14 columns, only one column contains float values while the other 13 contains integer values
*   Here the mean cholesterol level is around 247.350168;
         - Cholesterol level above 200 suggests that the heart health is in danger. Thus here the cholesterol level is higher.


*   Here the maximum heart rate achieved(thalach) is 149.599327;
         -The maximum heart rate(MHR) is calculated by subtracting the age from 220.
         - Here the mean age is 54.542088;
              - subtracting 54.542088 from 220 results in 165.457912, which is above the MHR of patients
                  -suggesting the patients heart rate is lower than needed




In [8]:
import numpy as np
import pandas as pd
import zipfile

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:
        data_table = pd.read_csv(file)

data_matrix = np.array(data_table)
print(data_matrix)


print(data_table.shape)

[[69.  1.  0. ...  1.  0.  0.]
 [69.  0.  0. ...  2.  0.  0.]
 [66.  0.  0. ...  0.  0.  0.]
 ...
 [35.  1.  3. ...  0.  2.  1.]
 [35.  0.  3. ...  0.  0.  0.]
 [35.  1.  3. ...  0.  2.  1.]]
(297, 14)


**matrix operations**

*   Here the matrix shape is 297,14; that means 297 rows and 14 columns

         - that is since it is not a square matrix,

               - we cannot find its;

                      -Determinant
                      -Inverse
                      -Eigen values and Eigen vectors
                      
              - But we can find it's;
                     -Rank
                     




In [9]:

rank = np.linalg.matrix_rank(data_matrix)
print(rank)

14


**Inference from Rank**



*   Here the Rank is 14;
         -columns are also 14;
                - which means there are 14 linearly independent columns are there and no column(here in this case data categories cannot be reduced)
                       - Thus all columns add meaning to the data
=



**Let's use a SINGLE FUNCTION**



*   From the previous project I have created a single wrapped function 'analyze_matrix' for analysing the key properties of matrix
   - Here we are using that single function

-check out the function here:


****

In [10]:
from matrix_module import analyze_matrix

zip_file_path = "archive.zip"
csv_filename_in_zip = "heart_cleveland_upload.csv"

with zipfile.ZipFile(zip_file_path, 'r') as zf:
    with zf.open(csv_filename_in_zip) as file:

        analyze_matrix(file)



     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     69    1   0       160   234    1        2      131      0      0.1   
1     69    0   0       140   239    0        0      151      0      1.8   
2     66    0   0       150   226    0        0      114      0      2.6   
3     65    1   0       138   282    1        2      174      0      1.4   
4     64    1   0       110   211    0        2      144      1      1.8   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
292   40    1   3       152   223    0        0      181      0      0.0   
293   39    1   3       118   219    0        0      140      0      1.2   
294   35    1   3       120   198    0        0      130      1      1.6   
295   35    0   3       138   183    0        0      182      0      1.4   
296   35    1   3       126   282    0        2      156      1      0.0   

     slope  ca  thal  condition  
0        1   1     0          0  
1        0   2     

In [53]:
#Let's do PCA analysis of this data
    #that is principal component analysis


import zipfile
import numpy as np
import pandas as pd

zip_file = "archive.zip"
csv_filename = "heart_cleveland_upload.csv"

with zipfile.ZipFile(zip_file, 'r') as zf:
  with zf.open(csv_filename) as file:
    data_table = pd.read_csv(file)

#let's do PCA
  #step 1 = conversion to matrix
  #step 2 = mean of each column
  #step 3 = subtracting the mean from each column and generating a mean centered matrix
  #step 4 = covariance of the matrix
  #step 5 = Eigen values and Eigen vectors
  #step 6 = ordering the eigen values from largest to smallest
  #step 7 = finding the eigen vector corresponding to the largest Eigen value
  #step 8 = calculating the PCA scores by multiplying mean centered matrix and eigen vectors of the largest eigen value

#step 1 = conversion to matrix;

data_matrix = data_table.select_dtypes(include='number').to_numpy()

#print(data_matrix)

#print(data_table)

#step 2: mean of each column
mean_values = []

for i in range(data_matrix.shape[1]):
  mean_value = np.mean(data_matrix[:,i])
  mean_values.append(mean_value)

mean_values = np.array(mean_values)
#print(mean_values)

#step 3: subtracting the mean from each column

mean_centered_matrix = data_matrix - mean_values
#print(mean_centered_matrix)

#step 4 : finding the covariance of the matrix

covariance_matrix = np.cov(mean_centered_matrix.T)
#print(covariance_matrix)

#step 5 : finding the Eigen values and Eigen vectors

eigen_values, eigen_vectors = np.linalg.eig(covariance_matrix)
#print(f"Eigen values: {eigen_values}",'\n')
#print(f"Eigen vectors: {eigen_vectors}")


#step 6 : sorting eigen values from largest to smallest

sorted_eigen_indices = np.argsort(eigen_values)[::-1]
sorted_eigen_values = eigen_values[sorted_eigen_indices]
#print(sorted_eigen_values)
#print(sorted_eigen_indices)

#How many largest eigen values are enough?
Total_variance = np.sum(eigen_values) #or sorted_eigen_values. It doesn't matter
variance_ratio = eigen_values/Total_variance
cumulative_variance = np.cumsum(variance_ratio)
#print(cumulative_variance)
#print(variance_ratio)

#here it provides 98% when we reach the third eigen eigen values. so we need to take upto 3 eigen values


#finding the corrsponding eigen vectors of the largest Eigen values, I have taken 3 vectors corresponding to three largest Eigen values

indices_of_large_eigenvalues = sorted_eigen_indices[0:3]

sorted_eigen_vectors = eigen_vectors[:,indices_of_large_eigenvalues]
#print(sorted_eigen_vectors)

#step 7 : calculating the PCA scores by multiplying mean centered matrix and eigen vectors of the largest eigen value

PCA_1_vector = sorted_eigen_vectors[:,0].reshape(-1,1)
PCA_2_vector = sorted_eigen_vectors[:,1].reshape(-1,1)
PCA_3_vector = sorted_eigen_vectors[:,2].reshape(-1,1)

# PCA_1_vector_T = PCA_1_vector.T # This transpose was causing the dimension mismatch
# PCA_2_vector_T = PCA_2_vector.T
# PCA_3_vector_T = PCA_3_vector.T

PCA_1_scores = mean_centered_matrix@PCA_1_vector
PCA_2_scores = mean_centered_matrix@PCA_2_vector
PCA_3_scores = mean_centered_matrix@PCA_3_vector

print(PCA_1_scores)
print(PCA_2_scores)
print(PCA_3_scores)

[[-1.12996295e+01]
 [-7.37513078e+00]
 [-1.98798147e+01]
 [ 3.52455753e+01]
 [-3.70291585e+01]
 [-1.80050292e+01]
 [-1.33261147e+01]
 [-1.29586999e+01]
 [-6.24194671e+00]
 [ 2.51554834e+01]
 [ 4.26787624e+01]
 [ 2.72602281e+01]
 [-4.30066948e+01]
 [ 3.66208128e+01]
 [-5.48108835e+01]
 [-6.21040251e+01]
 [ 5.14324484e+01]
 [-3.47080590e+01]
 [ 1.51754280e+01]
 [-3.02569995e+00]
 [-4.84281158e+01]
 [-1.75977705e+01]
 [-6.67340942e+01]
 [ 2.17860544e+01]
 [ 5.65767969e+01]
 [-5.13960547e-01]
 [ 5.86721827e-01]
 [-5.15632465e+01]
 [ 3.33512880e+01]
 [-3.91669707e+01]
 [-2.57410733e+01]
 [ 3.60844239e+01]
 [ 7.18544993e+01]
 [-1.13673606e+01]
 [ 1.33286522e+01]
 [-1.41104573e+01]
 [-7.92727991e+00]
 [ 4.70299950e+01]
 [-1.19355752e+01]
 [-2.63567695e+01]
 [ 2.80996788e+00]
 [ 1.45312137e+01]
 [ 9.44539958e+01]
 [ 6.02694425e+01]
 [ 4.05475037e+01]
 [ 3.85661128e+01]
 [ 7.67469812e+01]
 [-4.62545408e+01]
 [-4.26232299e+01]
 [-4.14174358e+00]
 [ 2.34837025e+01]
 [ 1.82701541e+01]
 [-1.9710521